relay-server/src/services/buffer/mod.rs

```diff
@@ -158,74 +165,88 @@ impl EnvelopeBufferService {
                 // once buffer V1 has been removed.
                 relay_log::trace!("EnvelopeBufferService push");
-                self.push(envelope).await;
+                self.push(buffer, envelope).await;
             }
             EnvelopeBuffer::NotReady(project_key, envelope) => {
                 relay_log::trace!("EnvelopeBufferService project not ready");
-                self.buffer.mark_ready(&project_key, false);
+                buffer.mark_ready(&project_key, false);
                 relay_statsd::metric!(counter(RelayCounters::BufferEnvelopesReturned) += 1);
-                self.push(envelope).await;
+                self.push(buffer, envelope).await;
             }
             EnvelopeBuffer::Ready(project_key) => {
                 relay_log::trace!("EnvelopeBufferService project ready {}", &project_key);
-                self.buffer.mark_ready(&project_key, true);
+                buffer.mark_ready(&project_key, true);
             }
         };
         self.changes = true;
     }
 
-    async fn push(&mut self, envelope: Box<Envelope>) {
-        if let Err(e) = self.buffer.push(envelope).await {
+    async fn push(&mut self, buffer: &mut PolymorphicEnvelopeBuffer, envelope: Box<Envelope>) {
+        if let Err(e) = buffer.push(envelope).await {
             relay_log::error!(
                 error = &e as &dyn std::error::Error,
                 "failed to push envelope"
             );
         }
     }
 
-    fn update_observable_state(&self) {
+    fn update_observable_state(&self, buffer: &mut PolymorphicEnvelopeBuffer) {
         self.has_capacity
-            .store(self.buffer.has_capacity(), Ordering::Relaxed);
+            .store(buffer.has_capacity(), Ordering::Relaxed);
     }
 }
 
 impl Service for EnvelopeBufferService {
     type Interface = EnvelopeBuffer;
 
     fn spawn_handler(mut self, mut rx: Receiver<Self::Interface>) {
+        let config = self.config.clone();
+        let memory_checker = self.memory_checker.clone();
         tokio::spawn(async move {
+            let mut buffer = PolymorphicEnvelopeBuffer::from_config(&config, memory_checker)
+                .await
+                .expect("Envelope buffer couldn't be initialized from the config");
+            buffer.initialize().await;
+
             relay_log::info!("EnvelopeBufferService start");
             loop {
                 relay_log::trace!("EnvelopeBufferService loop");
+
                 tokio::select! {
                     // NOTE: we do not select a bias here.
                     // On the one hand, we might want to prioritize dequeing over enqueing
                     // so we do not exceed the buffer capacity by starving the dequeue.
                     // on the other hand, prioritizing old messages violates the LIFO design.
                     () = self.wait_for_changes() => {
-                        if let Err(e) = self.try_pop().await {
+                        if let Err(e) = self.try_pop(&mut buffer).await {
```

Comment:
Instead of passing the buffer to every method, we could create a new helper type (`InnerService` or similar) that holds the buffer and other necessary fields, and make the loop a method of that new type.

Reply:
We could yes, idk if I like it though, seems like a lot of indirection for no real benefit. Already with our service framework, we have a bit of such indirections.